src/linter.h

```diff
@@ -18,39 +18,77 @@
 #define SRC_LINTER_H_
 
 #include <vector>
+#include <cstdio>
+#include <string>
 
 #include "absl/strings/string_view.h"
 #include "zetasql/base/status.h"
 #include "zetasql/parser/parse_tree_visitor.h"
+#include "zetasql/parser/parse_tree.h"
 #include "zetasql/parser/parser.h"
 #include "zetasql/public/parse_helpers.h"
+#include "zetasql/public/parse_resume_location.h"
 
 namespace zetasql {
 
 namespace linter {
 
-class KeywordExtractor {
+// It gets rule and applies that rule to every ASTnode it visit
+class RuleVisitor : public zetasql::NonRecursiveParseTreeVisitor {
  public:
-    KeywordExtractor(const zetasql::ASTNode *_node,
-                     std::vector<const zetasql::ASTNode *> *_keywords)
-        : node(_node),
-          keywords(_keywords) {}
-    void extract();
+    RuleVisitor(const std::function
+        <absl::Status(const zetasql::ASTNode*, absl::string_view)> &rule,
+                const absl::string_view &sql)
+        : rule_(rule), sql_(sql), result_(absl::OkStatus()) {}
+
+    zetasql_base::StatusOr<VisitResult> defaultVisit(
+        const ASTNode* node) override;
+
+    absl::Status GetResult() { return result_; }
 
  private:
-    const zetasql::ASTNode *node;
-    std::vector<const zetasql::ASTNode *> *keywords;
+    std::function
+        <absl::Status(const zetasql::ASTNode*, absl::string_view)> rule_;
+    absl::string_view sql_;
+    absl::Status result_;
 };
 
-// Checks whether given sql statement is a valid
-// GoogleSql statement
-absl::Status checkStatement(absl::string_view sql);
+//
+class ASTNodeRule {
+ public:
+    explicit ASTNodeRule(const std::function
+        <absl::Status(const zetasql::ASTNode*, absl::string_view)> rule)
+        : rule_(rule) {}
+    absl::Status ApplyTo(absl::string_view sql);
+ private:
+    std::function
+        <absl::Status(const zetasql::ASTNode*, absl::string_view)> rule_;
+};
+
+// Debugger that will be erased later.
+absl::Status PrintASTTree(absl::string_view sql);
 
 // Checks if the number of characters in any line
 // exceed a certain treshold
-absl::Status checkLineLength(absl::string_view sql, int lineLimit = 100,
+absl::Status CheckLineLength(absl::string_view sql, int lineLimit = 100,
     const char delimeter = '\n');
 
+// Checks whether given sql statement is a valid
+// GoogleSql statement
+absl::Status CheckStatement(absl::string_view sql);
```

Comment:
Don't you check now that the input contains multiple valid sql statements? You could just write
```
// Checks whether input can be parsed with ZetaSQL parser.
```

I also don't like the function name, it's ambiguous. CheckValidStatements? CheckParserSucceeds? CheckParseErrors?

Btw, do not use GoogleSQL anywhere in code or comments. It is ZetaSQL :D

Reply:
I am using CheckParserSucceeds, and it is ZetaSQL now 👍